x-pack/plugins/fleet/server/services/policy_preconfig.ts

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, SavedObjectsClientContract } from 'src/core/server';
+import { i18n } from '@kbn/i18n';
+import { groupBy } from 'lodash';
+
+import { packageToPackagePolicy } from '../../common';
+
+import type {
+  NewPackagePolicy,
+  AgentPolicy,
+  Installation,
+  Output,
+  InputsOverride,
+  FleetConfigType,
+} from '../../common';
+
+import { getPackageInfo } from './epm/packages';
+import { ensureInstalledPackage, isPackageInstalled } from './epm/packages/install';
+import { packagePolicyService } from './package_policy';
+import { agentPolicyService } from './agent_policy';
+
+export async function ensurePreconfiguredPackagesAndPolicies(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  policies: FleetConfigType['agentPolicies'] = [],
+  packages: FleetConfigType['packages'] = [],
+  defaultOutput: Output
+) {
+  // Validate configured packages to ensure there are no version conflicts
+  const packageNames = groupBy(packages, (pkg) => pkg.name);
+  const duplicatePackages = Object.entries(packageNames).filter(
+    ([, versions]) => versions.length > 1
+  );
+  if (duplicatePackages.length) {
+    // List duplicate packages as a comma-separated list of <package-name>:<semver>
+    // If there are multiple packages with duplicate versions, separate them with semicolons, e.g
+    // package-a:1.0.0, package-a:2.0.0; package-b:1.0.0, package-b:2.0.0
+    const duplicateList = duplicatePackages
+      .map(([, versions]) => versions.map((v) => `${v.name}:${v.version}`).join(', '))
+      .join('; ');
+
+    throw new Error(
+      i18n.translate('xpack.fleet.setup.duplicatePackageError', {
```

Comment:
can we make these i18n keys a bit more granular? like `xpack.fleet.preconfiguration.fooBarError`